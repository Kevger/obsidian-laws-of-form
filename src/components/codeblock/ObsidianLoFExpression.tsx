import LoF, { parser } from "laws-of-form-react";
import { ObsidianLoFBlockErrorMessage } from "./ObsidianLoFBlockErrorMessage";

/**
 * Displays a LoF expression based on the bracket notation e.g. ((a)b).
 *
 * If the expression is invalid, an error message is displayed instead.
 *
 * @param expression the LoF bracket expression to display
 * @returns JSX.Element with the LoF expression
 *
 */
export function ObsidianLoFExpression({
  expression,
}: {
  expression: string;
}): JSX.Element {
  try {
    // first we need to check if the expression is valid before creating the virtual dom
    parser.parse(expression, undefined);
  } catch (e) {
    return <ObsidianLoFBlockErrorMessage msg={e.toString()} />;
  }
  return <LoF>{expression}</LoF>;
}
