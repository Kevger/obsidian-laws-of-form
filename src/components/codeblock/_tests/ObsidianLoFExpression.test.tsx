import { create as render } from "react-test-renderer";
import { ObsidianLoFExpression } from "../ObsidianLoFExpression";

const PRIMITIVE_EXPRESSIONS = ["()", "( )", "", " "];

const NON_PRIMITIVE_EXPRESSIONS = [
  "(())",
  "(()) () ()",
  "(a)",
  "((hello world) ((abc)))",
  "()()()",
  "(((((((())))))))",
];

const REENTRY_EXPRESSIONS = [
  "((a[0])b$0)",
  "((a[0])(b[0])$0c)()",
  "([0]$0Autopoiesis)",
  "(((((((([0]a $1)$6[2])[1]$3)[0]$4)a$5)$2[6])a[5]) [4]$0)",
  "((a ((b [0] )c $1))(((d [2])(e $0))f[1]) g $2)",
];

describe("ObsidianLoFExpression", () => {
  beforeEach(() => {
    let randomMockValue = 0.000001;
    let dateMockValue = 1;
    jest.spyOn(global.Math, "random").mockImplementation(() => {
      randomMockValue = randomMockValue + 0.000001;
      return randomMockValue;
    });

    jest.spyOn(global.Date.prototype, "valueOf").mockImplementation(() => {
      return dateMockValue++;
    });
  });

  describe.each(PRIMITIVE_EXPRESSIONS)(
    "given we have an primitive expression %s",
    (primitiveExpression) => {
      it("renders correctly", () => {
        const component = render(
          <ObsidianLoFExpression expression={primitiveExpression} />
        );
        const snapshot = component.toJSON();
        expect(snapshot).toMatchSnapshot();
      });
    }
  );

  describe.each(NON_PRIMITIVE_EXPRESSIONS)(
    "given we have an non-primitive expression %s",
    (nonPrimitiveExpression) => {
      it("renders correctly", () => {
        const component = render(
          <ObsidianLoFExpression expression={nonPrimitiveExpression} />
        );
        const snapshot = component.toJSON();
        expect(snapshot).toMatchSnapshot();
      });
    }
  );

  describe.each(REENTRY_EXPRESSIONS)(
    "given we have an reentry expression %s",
    (reentryExpression) => {
      it("renders correctly", () => {
        const component = render(
          <ObsidianLoFExpression expression={reentryExpression} />
        );
        const snapshot = component.toJSON();
        expect(snapshot).toMatchSnapshot();
      });
    }
  );

  describe("given we have an invalid expression", () => {
    it("should return an ObsidianLoFBlockErrorMessage in virtual dom", () => {
      const component = render(<ObsidianLoFExpression expression="(()" />);
      expect((component.toTree()!.rendered as any).type.name).toBe(
        "ObsidianLoFBlockErrorMessage"
      );
    });

    it("should contain the error message", () => {
      const component = render(<ObsidianLoFExpression expression="(()" />);
      expect((component.toTree()!.rendered as any).rendered.rendered[0]).toBe(
        'SyntaxError: Expected Expression, end of input, or whitespace but "(" found.'
      );
    });
  });
});
