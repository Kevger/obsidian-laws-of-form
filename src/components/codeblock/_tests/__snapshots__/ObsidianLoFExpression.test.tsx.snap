// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObsidianLoFExpression given we have an non-primitive expression (((((((()))))))) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_8"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_7"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc_6"
      >
        <div
          className="cross "
          id="Cross_1_0003cxkbtc_5"
        >
          <div
            className="cross "
            id="Cross_1_0003cxkbtc_4"
          >
            <div
              className="cross "
              id="Cross_1_0003cxkbtc_3"
            >
              <div
                className="cross "
                id="Cross_1_0003cxkbtc_2"
              >
                <div
                  className="cross "
                  id="Cross_1_0003cxkbtc_1"
                />
                 
              </div>
               
            </div>
             
          </div>
           
        </div>
         
      </div>
       
    </div>
     
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an non-primitive expression (()) () () renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_2"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    />
     
  </div>,
  " ",
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_3"
  />,
  " ",
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_4"
  />,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an non-primitive expression (()) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_2"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    />
     
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an non-primitive expression ((hello world) ((abc))) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_4"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    >
      hello world
    </div>
     
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_3"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc_2"
      >
        abc
      </div>
       
    </div>
     
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an non-primitive expression ()()() renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_1"
  />,
  " ",
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_2"
  />,
  " ",
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_3"
  />,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an non-primitive expression (a) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_1"
  >
    a
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an primitive expression   renders correctly 1`] = `" "`;

exports[`ObsidianLoFExpression given we have an primitive expression  renders correctly 1`] = `" "`;

exports[`ObsidianLoFExpression given we have an primitive expression ( ) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_1"
  />,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an primitive expression () renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_1"
  />,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an reentry expression (((((((([0]a $1)$6[2])[1]$3)[0]$4)a$5)$2[6])a[5]) [4]$0) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc0"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_7"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc2"
      >
        <div
          className="cross "
          id="Cross_1_0003cxkbtc5"
        >
          <div
            className="cross "
            id="Cross_1_0003cxkbtc4"
          >
            <div
              className="cross "
              id="Cross_1_0003cxkbtc3"
            >
              <div
                className="cross "
                id="Cross_1_0003cxkbtc6"
              >
                <div
                  className="cross "
                  id="Cross_1_0003cxkbtc1"
                >
                  a
                </div>
                 
              </div>
               
            </div>
             
          </div>
           
          a
        </div>
         
      </div>
       
      a
    </div>
     
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an reentry expression ((a ((b [0] )c $1))(((d [2])(e $0))f[1]) g $2) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc2"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_3"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc1"
      >
        <div
          className="cross "
          id="Cross_1_0003cxkbtc_1"
        >
          b
        </div>
         
        c
      </div>
       
      a
    </div>
     
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_7"
    >
      <div
        className="cross "
        id="Cross_1_0003cxkbtc_6"
      >
        <div
          className="cross "
          id="Cross_1_0003cxkbtc_4"
        >
          d
        </div>
         
        <div
          className="cross "
          id="Cross_1_0003cxkbtc0"
        >
          e
        </div>
         
      </div>
       
      f
    </div>
     
    g
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an reentry expression ((a[0])(b[0])$0c)() renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc0"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    >
      a
    </div>
     
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_2"
    >
      b
    </div>
     
    c
  </div>,
  " ",
  <div
    className="cross "
    id="Cross_1_0003cxkbtc_4"
  />,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an reentry expression ((a[0])b$0) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc0"
  >
    <div
      className="cross "
      id="Cross_1_0003cxkbtc_1"
    >
      a
    </div>
     
    b
  </div>,
  " ",
  " ",
]
`;

exports[`ObsidianLoFExpression given we have an reentry expression ([0]$0Autopoiesis) renders correctly 1`] = `
[
  <div
    className="cross "
    id="Cross_1_0003cxkbtc0"
  >
    Autopoiesis
  </div>,
  " ",
  " ",
]
`;
